### PR TITLE
[ADD] sale: add pdf quote builder

### PR DESCRIPTION
**Pull request by Demesmaeker**
The value provided by the sale quotation builder feature is small due to many problems, and the
technical debt behind is high. For instance:

- The feature is nice but hidden: not visible on quotation, the most visible part is the button
"Design Template" on the quotation template (if you did activate the feature), it requires having
eCommerce installed...

- The feature relies on fields that are shared with eCommerce which leads to unwanted behavior:
e.g. updating the bottom of the product page in eCommerce will impact quotation html description on
the portal

- The feature works as template, but it is unclear for users: updating the HTML description on a
specific quotation will only impact that quotation. Updating the HTML page of a quotation only
update this quotation

- The feature is technically complex

As we want to keep most of the value of the feature but reduce its complexity, we remove the
quotation builder feature that'll get replace with the PDF quote builder feature.

When enabled, this new feature adds the possibility to set a PDF header, footer and some product
documents to the quotation report.

These PDF can contain forms that'll then be filled using Odoo database.

This replaces the previous sale quotation builder feature.

task-3249142

See also:

- https://github.com/odoo/documentation/pull/5863
- https://github.com/odoo/upgrade/pull/5168

### CODE REVIEW

File: addons/sale_pdf_quote_builder/models/sale_order_template.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class SaleOrderTemplate(models.Model):
+    _inherit = 'sale.order.template'
+
+    header = fields.Binary()
+    footer = fields.Binary()
```
Comment (Feyensv) [inline addons/sale_pdf_quote_builder/models/sale_order_template.py:10]: I think it should default to company header & footer ;)

File: addons/sale_pdf_quote_builder/data/sale_pdf_quote_builder_demo.xml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
```
Comment (Feyensv) [inline addons/sale_pdf_quote_builder/data/sale_pdf_quote_builder_demo.xml:3]:
```suggestion
```

File: addons/sale_pdf_quote_builder/models/res_company.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+from odoo import api, fields, models
+
+from odoo.addons.sale_pdf_quote_builder import utils as pdf_utils
+
+
+class ResCompany(models.Model):
+    _inherit = 'res.company'
+
+    sale_header = fields.Binary('Quotation PDF Header')
+    sale_footer = fields.Binary('Quotation PDF Footer')
```
Comment (Feyensv) [inline addons/sale_pdf_quote_builder/models/res_company.py:12]:
```suggestion
    sale_header = fields.Binary(string="Quotation PDF Header")
    sale_footer = fields.Binary(string="Quotation PDF Footer")
```

**Review by Feyensv — COMMENTED** [review]
:duck:

File: addons/sale_pdf_quote_builder/utils.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+def check_base64_is_pdf(base64):
+    """ Check if the given base64 is a valid PDF document.
+
+    :param bytes base64: The base64 bytes to check
+    :return: True if the string is a valid PDF document
+    :rtype: bool
+    """
+    if base64 and not str(base64).startswith("b'JVBERi0"):
+        raise ValidationError(_("Only PDF documents can be used as header or footer."))
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/sale_pdf_quote_builder/utils.py:15]: Old comment? Already done =]

File: addons/sale_pdf_quote_builder/models/ir_actions_report.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                footer = record.sale_footer
+
+                included_product_docs = self.env['product.document']
+                for line in order.order_line:
+                    document = (
+                        line.product_id.product_document_ids
+                        | line.product_template_id.product_document_ids
+                    )
+                    doc_to_include = document.filtered(lambda d: d.attached_on == 'inside')
+                    included_product_docs = included_product_docs | doc_to_include
```
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/sale_pdf_quote_builder/models/ir_actions_report.py:38]: Indeed, my bad >.<

File: addons/sale_pdf_quote_builder/views/sale_order_template_views.xml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+                                Set a PDF header and footer to produce a high quality quotation,
+                                built by concatenating the template header, product sheets, the
+                                quotation with product and prices, and the footer.
+                            </p><p class="text-muted" colspan="2">
+                                To include product sheets in your Quotation templates, add documents
+                                on product forms, and set their visibility as "Inside Quotes".
+                            </p>
+                        </group>
+                    </group>
+                </page>
```
Comment (dalaOdoo) [inline addons/sale_pdf_quote_builder/views/sale_order_template_views.xml:27]:
```suggestion
                <page name="pdf_quote" string="PDF Quote Builder">
                    <group>
                        <p class="text-muted" colspan="2">
                            Provide header pages and footer pages to compose an attractive quotation
                            with more information about your company, your products and your services. 
                            The pdf of your quotes will be built by putting together header pages, 
                            product descriptions, details of the quote and then the footer pages.
                        </p>
                        <group>
                            <field name="sale_header" options="{'accepted_file_extensions': '.pdf'}"/>
                            <field name="sale_footer" options="{'accepted_file_extensions': '.pdf'}"/>
                        </group>
                        <p class="text-muted" colspan="2">
                            Products descriptions are pdf documents you can add directly on products. 
                            To do so, go on a product, find the "product documents" button, then add a 
                            new pdf document with a visibility set as "Inside Quotes". For each product 
                            in the quote, if the product has an "inside quotes" document, this document
                            will be added after header pages and before the quotation details.
                        </p>
                        <p class="text-muted" colspan="2">
                            Some information specific to the quote (customer name, quotation reference, ... ) 
                            can be injected in these documents using pdf forms.
                            Refer to the documentation to know more about this feature.  
                        </p>
                    </group>
                </page>
```

File: addons/sale_pdf_quote_builder/data/sale_pdf_quote_builder_demo.xml [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
```
Comment (Feyensv) [inline addons/sale_pdf_quote_builder/data/sale_pdf_quote_builder_demo.xml:2]: noupdate ?

**Comment by Feyensv** [comment]
@robodoo override=ci/style

**Comment by Feyensv** [comment]
@Xavier-Do can we have a little override of ci/style here please ? :pleading_face: 
The files have been compressed already and we want the demo files to be qualitative enough, even if they are heavier than 100KB.

**Comment by Xavier-Do** [comment]
@robodoo override=ci/style

**Comment by Feyensv** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**